docs/proposals/20201019-prometheus-rules-in-observatorium-api.md

```diff
@@ -0,0 +1,238 @@
+4. Object storage returns success to the Rule Backend.
+5. Rule backend returns success to the API.
+6. API returns success to the Tenant.
+
+#### 2. Sync Rules to Rule
+1. Periodically, thanos-ruler-syncer queries the API for all rules for all tenants.
+   * Do we need to expose an endpoint to return all of the rules? Unclear at the moment.
+   * NB: We can definitely make some optimisations here. Suggestions welcome.
+2. API requests data from the rule storage backend.
+3. Rule storage backend requests data from Object storage.
+4. Object storage returns data to the Rule storage backend.
+5. Rule storage backend returns data to the API.
+6. API returns data to thanos-ruler-syncer.
+7. Thanos-ruler-syncer combines all of the data into one file, and writes this file to a directory that is shared with Thanos Rule.
+8. Thanos-ruler-syncer calls Thanos Rule's `/-/reload` endpoint to reload newly created configuration.
+9. Thanos Rule begins to evaluate the recording rules and stores the results in its local TSDB instace.
+
+#### 3. Recording Rule Query
+
+1. Tenant makes a query request to the API containing a recording rule metric
+2. API performs authentication and authorization of the request then queries the data from Thanos Query.
+3. Query has Rule configured as one of its Stores, and thus requests the data from Rule.
+4. Rule returns the metric value of the recording rule to Query.
+5. Query returns the result to API.
+6. API returns the result to the tenant.
+
+## Alternatives
+
+### More frequent deploys
+
+The above problem has manifested because our deploys are too infrequent for our tenants. Could we not use this as an opportunity to optimize the velocity at which we deploy to production? That way we have the benefit of solving the near-term problems of our tenants, and also rolling things out more frequently.
+
+As it stands, any rollout to production requires coordination and management from a member of the operating team. As the number of tenants and tenants becomes larger, the operating team becomes more of a bottleneck to common tenant operations.
+
+Our goal here is to remove our team entirely from the control loop of updating Prometheus rules. We are trading off immediate term technical complexity against longer-term scalability and happier multi-tenancy operation.
+
+This option also does not address the trust-boundary issue identified above.
+
+### Per-tenant GitOps
+
+Another option would be for each tenant to commit their rule configuration into a git repository they control, and then we synchronise that into the right place in our infrastructure based on a URL they provide. This would eliminate the need for us to manage state on their behalf.
+
+This is a neat idea that avoids deploy dependencies, but has some drawbacks:
```

Comment:
Optional: I think in general having to manage additional git repo sounds already like enough of an overhead for the tenants, even without validation / secrets, so I find our actual proposal much more 'tenant-friendly'.

Reply:
Good point :) Added that in the list